src/backend/distributed/planner/merge_planner.c

```diff
@@ -642,86 +1030,201 @@ MergeQuerySupported(Oid resultRelationId, Query *originalQuery, bool multiShardQ
 			RaiseDeferredError(deferredError, ERROR);
 		}
 	}
+}
 
-	deferredError =
-		InsertDistributionColumnMatchesSource(resultRelationId, originalQuery);
-	if (deferredError)
-	{
-		/* MERGE's unsupported scenario, raise the exception */
-		RaiseDeferredError(deferredError, ERROR);
-	}
 
-	if (multiShardQuery)
+/*
+ * MergeCommandResultIdPrefix returns the prefix to use for intermediate results of
+ * an MERGE INTO ... USING source-query results via the coordinator.
+ */
+static char *
+MergeCommandResultIdPrefix(uint64 planId)
+{
+	StringInfo resultIdPrefix = makeStringInfo();
+	appendStringInfo(resultIdPrefix, "merge_into_" UINT64_FORMAT, planId);
+	return resultIdPrefix->data;
+}
+
+
+/*
+ * ValidateAndReturnVarIfSupported Checks for valid expressions of type Var, and
+ * returns the Var if it finds one, for everything else, raises an exception.
+ */
+static Var *
+ValidateAndReturnVarIfSupported(Expr *entryExpr)
+{
+	switch (entryExpr->type)
 	{
-		deferredError =
-			DeferErrorIfUnsupportedSubqueryPushdown(originalQuery,
-													plannerRestrictionContext);
-		if (deferredError)
+		case T_Var:
+		{
+			/* Found a Var inserting into target's distribution column */
+			return (Var *) entryExpr;
+		}
+
+		case T_Const:
+		{
+			/* Inserting constant VALUES is not allowed */
+		}
+
+		case T_FuncExpr:
+		{
+			FuncExpr *funcExpr = (FuncExpr *) entryExpr;
+
+			if (funcExpr->funcformat == COERCE_IMPLICIT_CAST &&
```

Comment:
as noted in another place, using `strip_implicit_coercions` might be better as it covers a lot more valid cases than this.

Basically, before calling `ValidateAndReturnVarIfSupported` we should do `strip_implicit_coercions(copyObject(expr))`.

I guess we need copy anyway as we are modifying the original query?

Reply:
yeah, makes sense. Removed processing `FuncExpr` and let `strip_implicit_coercions` do the job